### PR TITLE
Improve theory combination in the presence of real algebraic numbers

### PR DESCRIPTION
**Pull request by nafur**
This PR changes how we handle real algebraic numbers in theory combination and model construction.
The goal is to improve `getEqualityStatus()` and produce proper models more often.
We now use a RAN-aware evaluator for `getEqualityStatus()` and change the way how the nonlinear extension finalizes its model.

### CODE REVIEW

**Comment by nafur** [comment]
The changes have no significant impact on benchmarks:
```
$ cmpr.py -G slurmwd/01051455-cvc5-*
                                   status  total  solved   sat  unsat   best  timeout  memout  error  uniq  disagr  time_cpu   memory
benchmark   config                                                                                                                   
AUFDTNIRA   01051455-cvc5-eqstatus     ok    471     328     0    328    307       59       0      0     0       0   17909.5  16421.1
            01051455-cvc5-master       ok    471     328     0    328    302       59       0      0     0       0   17910.4  16673.5
AUFFPDTNIRA 01051455-cvc5-eqstatus     ok     13       7     0      7      7        6       0      0     0       0    1805.5    468.5
            01051455-cvc5-master       ok     13       7     0      7      5        6       0      0     0       0    1805.7    459.8
AUFNIRA     01051455-cvc5-eqstatus     ok   1480    1041     0   1041   1022      390       0      0     1       0  118028.6 184368.9
            01051455-cvc5-master       ok   1480    1040     0   1040   1022      391       0      0     0       0  118587.6 200647.7
NRA         01051455-cvc5-eqstatus     ok   4058    4030   221   3809   4022       28       0      0     0       0    8463.6   8134.0
            01051455-cvc5-master       ok   4058    4030   221   3809   4022       28       0      0     0       0    8463.8   8097.4
QF_NIRA     01051455-cvc5-eqstatus     ok      3       2     0      2      2        1       0      0     1       0     594.5   2817.3
            01051455-cvc5-master       ok      3       1     0      1      1        2       0      0     0       0     601.6   2624.9
QF_NRA      01051455-cvc5-eqstatus     ok  11552   10714  5093   5621  10171      837       1      0     2       0  287454.4 363232.4
            01051455-cvc5-master       ok  11552   10713  5093   5620   9965      838       1      0     1       0  287699.0 363290.5
QF_UFNRA    01051455-cvc5-eqstatus     ok     37      22    11     11     16       15       0      0     0       0    5659.2   2159.0
            01051455-cvc5-master       ok     37      22    11     11     17       15       0      0     0       0    5657.3   1806.0
UFDTNIRA    01051455-cvc5-eqstatus     ok   1842    1623     0   1623   1539       48       0      0     0       0   16807.4  29408.6
            01051455-cvc5-master       ok   1842    1623     0   1623   1529       48       0      0     0       0   16825.5  31968.9
UFFPDTNIRA  01051455-cvc5-eqstatus     ok     21       7     0      7      7        8       0      0     0       0    2411.2   2785.7
            01051455-cvc5-master       ok     21       7     0      7      7        8       0      0     0       0    2411.3   3018.0
```
However, they make the implementation (more) correct, as witnessed by the regression.

File: src/theory/arith/arith_evaluator.cpp [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+#include "theory/theory.h"
+#include "util/real_algebraic_number.h"
+
+namespace cvc5 {
+namespace theory {
+namespace arith {
+
+namespace {
+
+RealAlgebraicNumber evaluate(TNode expr,
```
Comment (ajreynol) [inline src/theory/arith/arith_evaluator.cpp:14]: Shouldn't we do a dag traversal for this? 
Also technically we should make this iterative not recursive.

File: src/theory/arith/arith_evaluator.h [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+#define CVC5__THEORY__ARITH__ARITH_EVALUATOR_H
+
+#include "expr/node.h"
+#include "smt/env.h"
+
+namespace cvc5 {
+namespace theory {
+namespace arith {
+
+bool isExpressionZero(Env& env, Node expr, const std::map<Node, Node>& model);
```
Comment (ajreynol) [inline src/theory/arith/arith_evaluator.h:14]: Document.

File: src/theory/arith/nl/cad_solver.h [hunk 3]
```diff
@@ -82,6 +82,8 @@ class CadSolver: protected EnvObj
   bool constructModelIfAvailable(std::vector<Node>& assertions);
 
  private:
+  void addToModel(TNode var, TNode value) const;
```
Comment (ajreynol) [inline src/theory/arith/nl/cad_solver.h:85]: Add a brief doc on this.

File: src/theory/arith/nl/nonlinear_extension.h [hunk 4]
```diff
@@ -111,10 +111,8 @@ class NonlinearExtension : EnvObj
   void checkFullEffort(std::map<Node, Node>& arithModel,
                        const std::set<Node>& termSet);
 
-  /**
-   * Finalize the given model by adding approximations and witnesses.
-   */
-  void finalizeModel(TheoryModel* tm);
+  Node getModelValue(TNode var) const;
```
Comment (ajreynol) [inline src/theory/arith/nl/nonlinear_extension.h:114]: Document these.

**Review by ajreynol — CHANGES_REQUESTED** [review]

File: src/theory/arith/arith_evaluator.cpp (hunk 1 above)
Comment (nafur) [reply src/theory/arith/arith_evaluator.cpp:14]: Theoretically, yes. However, we will soon replace this code by properly integrating RANs into the arithmetic rewriter...

Comment (ajreynol) [reply src/theory/arith/arith_evaluator.cpp:14]: Ok thats fine for now then.

**Review by ajreynol — APPROVED** [review]

**Pull request merged**